File: src/translv.f90
SUBROUTINE translv

!-----------------------------------------------------------------------
!
! Solution driver. Contains the time and outer loops. Calls for outer
! iteration work. Checks convergence and handles eventual output.
!
!-----------------------------------------------------------------------

  USE global_module, ONLY: i_knd, r_knd, ounit, zero, half, one, two

  USE plib_module, ONLY: glmax, comm_snap, iproc, root, thread_num,    &
    ichunk, nthreads, nnested

  USE geom_module, ONLY: geom_alloc, geom_dealloc, dinv, param_calc,   &
    nx, ny_gl, nz_gl

  USE sn_module, ONLY: nang, noct, mu, eta, xi, nmom

  USE data_module, ONLY: ng, v, vdelt, mat, sigt, siga, slgg, src_opt, &
    qim

  USE control_module, ONLY: nsteps, timedep, dt, oitm, otrdone,        &
    control_alloc, control_dealloc, dfmxo, it_det, popout, swp_typ

  USE utils_module, ONLY: print_error, stop_run

  USE solvar_module, ONLY: solvar_alloc, ptr_in, ptr_out, t_xs, a_xs,  &
    s_xs, flux0, fluxm

  USE expxs_module, ONLY: expxs_reg

  USE outer_module, ONLY: outer

  USE time_module, ONLY: tslv, wtime, tgrind, tparam

  USE analyze_module, ONLY: pop_calc

  IMPLICIT NONE
!_______________________________________________________________________
!
! Local variables
!_______________________________________________________________________

  CHARACTER(LEN=1) :: star='*'

  CHARACTER(LEN=64) :: error

  INTEGER(i_knd) :: cy, otno, ierr, g, l, i, tot_iits, cy_iits, out_iits

  REAL(r_knd) :: sf, time, t1, t2, t3, t4, t5, tmp

  REAL(r_knd), DIMENSION(:,:,:,:,:,:), POINTER :: ptr_tmp
!_______________________________________________________________________
!
! Call for data allocations. Some allocations depend on the problem
! type being requested.
!_______________________________________________________________________

  CALL wtime ( t1 )

  ierr = 0
  error = ' '

  CALL geom_alloc ( nang, ng, swp_typ, ichunk, ierr )
  CALL glmax ( ierr, comm_snap )
  IF ( ierr /= 0 ) THEN
    error = '***ERROR: GEOM_ALLOC: Allocation error of sweep parameters'
    CALL print_error ( ounit, error )
    CALL stop_run ( 1, 3, 0, 0 )
  END IF

  CALL solvar_alloc ( ierr )
  CALL glmax ( ierr, comm_snap )
  IF ( ierr /= 0 ) THEN
    error = '***ERROR: SOLVAR_ALLOC: Allocation error of solution ' // &
            'arrays'
    CALL print_error ( ounit, error )
    CALL stop_run ( 1, 3, 1, 0 )
  END IF

  CALL control_alloc ( ng, ierr )
  CALL glmax ( ierr, comm_snap )
  IF ( ierr /= 0 ) THEN
    error = '***ERROR: CONTROL_ALLOC: Allocation error of control ' // &
            'arrays'
    CALL print_error ( ounit, error )
    CALL stop_run ( 1, 3, 2, 0 )
  END IF

  CALL wtime ( t2 )
  tparam = tparam + t2 - t1
!_______________________________________________________________________
!
! The time loop solves the problem for nsteps. If static, there is
! only one step, and it does not have any time-absorption or -source
! terms. Set the pointers to angular flux arrays. Set time to one for
! static for proper multiplication in octsweep.
!_______________________________________________________________________

  IF ( iproc == root ) THEN
    WRITE( *, 201)     ( star, i = 1, 80 )
    WRITE( ounit, 201) ( star, i = 1, 80 )
  END IF

  tot_iits = 0

  time_loop: DO cy = 1, nsteps

    CALL wtime ( t3 )

    vdelt = zero
    time = one
    IF ( timedep == 1 ) THEN
      IF ( iproc == root ) THEN
        WRITE( *, 202 )     ( star, i = 1, 30 ), cy
        WRITE( ounit, 202 ) ( star, i = 1, 30 ), cy
      END IF
      vdelt = two / ( dt * v )
      time = dt * ( REAL( cy, r_knd ) - half )
    END IF

    IF ( cy > 1 ) THEN
      ptr_tmp => ptr_out
      ptr_out => ptr_in
      ptr_in  => ptr_tmp
    END IF
!_______________________________________________________________________
!
!   Prepare some cross sections: total, in-group scattering, absorption.
!   Keep in the time loop for better consistency with PARTISN. Set up
!   geometric sweep parameters. Reset flux moments to zero at start of
!   each time step. Parallelize group loop with threads.
!_______________________________________________________________________

  !$OMP PARALLEL DO SCHEDULE(STATIC,1) NUM_THREADS(nthreads*nnested)   &
  !$OMP& DEFAULT(SHARED) PRIVATE(g,l)
    DO g = 1, ng
      CALL expxs_reg ( siga(:,g), mat, a_xs(:,:,:,g) )
      CALL expxs_reg ( sigt(:,g), mat, t_xs(:,:,:,g) )
      CALL param_calc ( nang, ichunk, mu, eta, xi, t_xs(:,:,:,g),      &
        vdelt(g), dinv(:,:,:,:,:,g) )
      DO l = 1, nmom
        CALL expxs_reg ( slgg(:,l,g,g), mat, s_xs(:,:,:,l,g) )
      END DO
      flux0(:,:,:,g) = zero
      fluxm(:,:,:,:,g) = zero
    END DO
  !$OMP END PARALLEL DO
!_______________________________________________________________________
!
!   Scale the manufactured source for time
!_______________________________________________________________________

    IF ( src_opt == 3 ) THEN
      IF ( cy == 1 ) THEN
        qim = time*qim
      ELSE
        sf = REAL( 2*cy - 1, r_knd ) / REAL( 2*cy-3, r_knd )
        qim = qim*sf
      END IF
    END IF
!_______________________________________________________________________
!
!   Using Jacobi iterations in energy, and the work in the outer loop
!   will be parallelized with threads.
!_______________________________________________________________________

    otrdone = .FALSE.

    cy_iits = 0

    IF ( iproc==root .AND. it_det==0 ) THEN
      WRITE( *, 203 )
      WRITE( ounit, 203 )
    END IF

    CALL wtime ( t4 )
    tparam = tparam + t4 - t3

    outer_loop: DO otno = 1, oitm

      IF ( iproc==root .AND. it_det==1 ) THEN
        WRITE( *, 204 )     ( star, i = 1, 20 ), otno
        WRITE( ounit, 204 ) ( star, i = 1, 20 ), otno
      END IF
!_______________________________________________________________________
!
!     Perform an outer iteration. Add up inners. Check convergence.
!_______________________________________________________________________

      CALL outer ( out_iits )

      cy_iits = cy_iits + out_iits

      IF ( iproc == root ) THEN
        WRITE( *, 205 )     otno, dfmxo, out_iits
        WRITE( ounit, 205 ) otno, dfmxo, out_iits
      END IF

      IF ( otrdone ) EXIT outer_loop

    END DO outer_loop
!_______________________________________________________________________
!
!   Compute and print the particle spectrum every cycle.
!_______________________________________________________________________

    IF ( popout == 2 ) CALL pop_calc ( cy, time )
!_______________________________________________________________________
!
!   Print the time cycle details. Add time cycle iterations.
!_______________________________________________________________________

    IF ( iproc == root ) THEN
      IF ( timedep == 1 ) THEN
        IF ( otrdone ) THEN
          WRITE( *, 206 )     cy, time, otno, cy_iits
          WRITE( ounit, 206 ) cy, time, otno, cy_iits
        ELSE
          WRITE( *, 207 )     cy, time, otno, cy_iits
          WRITE( ounit, 207 ) cy, time, otno, cy_iits
        END IF
      ELSE
        IF ( otrdone ) THEN
          WRITE( *, 208 )     otno, cy_iits
          WRITE( ounit, 208 ) otno, cy_iits
        ELSE
          WRITE( *, 209 )     otno, cy_iits
          WRITE( ounit, 209 ) otno, cy_iits
        END IF
      END IF
    END IF

    tot_iits = tot_iits + cy_iits

  END DO time_loop
!_______________________________________________________________________
!
!   Compute and print the particle spectrum only at end of calc.
!_______________________________________________________________________

    IF ( popout == 1 ) CALL pop_calc ( cy, time )
!_______________________________________________________________________
!
!   Final prints.
!_______________________________________________________________________

  IF ( iproc == root ) THEN
    IF ( timedep == 1 ) THEN
      WRITE( *, 210 )     ( star, i = 1, 30 ), tot_iits
      WRITE( ounit, 210 ) ( star, i = 1, 30 ), tot_iits
    END IF
    WRITE( *, 211 )     ( star, i = 1, 80 )
    WRITE( ounit, 211 ) ( star, i = 1, 80 )
  END IF

  CALL wtime ( t5 )
  tslv = t5 - t1
  tmp = REAL( nx, r_knd ) * REAL( ny_gl, r_knd ) * REAL( nz_gl, r_knd )&
        * REAL( nang, r_knd ) * REAL( noct, r_knd )                    &
        * REAL( tot_iits, r_knd )
  tgrind = tslv*1.0E9_r_knd / tmp
!_______________________________________________________________________

  201 FORMAT( 10X, 'keyword Iteration Monitor', /, 80A )
  202 FORMAT( /, 1X, 30A, /, 2X, 'Time Cycle ', I3 )
  203 FORMAT( 2X, 'Outer' )
  204 FORMAT( 1X, 20A, /, 2X, 'Outer ', I3 )
  205 FORMAT( 2X, I3, 4X, 'Dfmxo=', ES11.4, 4X, 'No. Inners=', I5 )
  206 FORMAT( /, 2X, 'Cycle=', I4, 4X, 'Time=', ES11.4, 4X, 'No. ',    &
              'Outers=', I4, 4X, 'No. Inners=', I5 )
  207 FORMAT( /, 2X, '*WARNING: Unconverged outer iterations', /, 2X,  &
             'Cycle=', I4, 4X, 'Time=', ES11.4, 4X, 'No. Outers=', I4, &
             4X, 'No. Inners=', I5, / )
  208 FORMAT( /, 2X, 'No. Outers=', I4, 4X, 'No. Inners=', I5 )
  209 FORMAT( /, 2X, '*WARNING: Unconverged outer iteration!', /, 2X,  &
              'No. Outers=', I4, 4X, 'No. Inners=', I5, / )
  210 FORMAT( /, 1X, 30A, /, 2X, 'Total inners for all time steps, '   &
              'outers = ', I6 )
  211 FORMAT( /, 80A, / )
!_______________________________________________________________________
!_______________________________________________________________________

END SUBROUTINE translv

File: ports/snap-c/translv.c
/***********************************************************************
 * Solution driver. Contains the time and outer loops. Calls for outer
 * iteration work. Checks convergence and handles eventual output.
 ***********************************************************************/
#include "snap.h"

void translv ( input_data *input_vars, para_data *para_vars, time_data *time_vars,
               geom_data *geom_vars, sn_data *sn_vars, data_data *data_vars,
               control_data *control_vars, solvar_data *solvar_vars, mms_data *mms_vars,
               sweep_data *sweep_vars, dim_sweep_data *dim_sweep_vars,
               FILE *fp_out, int *ierr, char **error )
{
/***********************************************************************
 * Local variables
 ***********************************************************************/
    int cy, otno, i, tot_iits, cy_iits, out_iits;

    int tmpStrLen = 0;

    int ang, x, y, z, oct, g; // indexes for 6D array

    double sf, time, t1, t2, t3, t4, t5, t6, t7, tmp;

    double *ptr_tmp;

    *ierr = 0;

/***********************************************************************
 * Call for data allocations. Some allocations depend on the problem
 * type being requested.
 ***********************************************************************/
    t1 = wtime ();

    geom_alloc ( input_vars, geom_vars, ierr );

    glmax_i ( ierr, COMM_SNAP );

    if ( *ierr != 0 )
    {
        tmpStrLen = strlen ("   ***ERROR: GEOM_ALLOC:"
                            " Allocation error of sweep parameters.\n");
        ALLOC_STR( *error, tmpStrLen + 1, ierr);
        snprintf ( (char *) *error, tmpStrLen + 1,
                   "   ***ERROR: GEOM_ALLOC:"
                   " Allocation error of sweep parameters.\n" );

        print_error ( fp_out, *error, IPROC, ROOT );

        stop_run ( 3, 0, 0, para_vars, sn_vars, data_vars, mms_vars,
                   geom_vars, solvar_vars, control_vars );
    }

    solvar_alloc ( input_vars, sn_vars, solvar_vars, ierr );

    glmax_i ( ierr, COMM_SNAP );

    if ( *ierr != 0 )
    {
        tmpStrLen = strlen ("   ***ERROR: SOLVAR_ALLOC:"
                            " Allocation error of solution.\n");
        ALLOC_STR( *error, tmpStrLen + 1, ierr);
        snprintf ( (char *) *error, tmpStrLen + 1,
                   "   ***ERROR: SOLVAR_ALLOC:"
                   " Allocation error of solution.\n" );

        print_error ( fp_out, *error, IPROC, ROOT );

        stop_run ( 3, 1, 0, para_vars, sn_vars, data_vars, mms_vars,
                   geom_vars, solvar_vars, control_vars );
    }

    control_alloc ( input_vars, control_vars, ierr );

    glmax_i ( ierr, COMM_SNAP );

    if ( *ierr != 0 )
    {
        tmpStrLen = strlen ("   ***ERROR: CONTROL_ALLOC:"
                            " Allocation error of control.\n");
        ALLOC_STR( *error, tmpStrLen + 1, ierr);
        snprintf ( (char *) *error, tmpStrLen + 1,
                   "   ***ERROR: CONTROL_ALLOC:"
                   " Allocation error of control.\n" );

        print_error ( fp_out, *error, IPROC, ROOT );

        stop_run ( 3, 2, 0, para_vars, sn_vars, data_vars, mms_vars,
                   geom_vars, solvar_vars, control_vars );
    }

/***********************************************************************
 * Call for setup of the mini-KBA diagonal map
 ***********************************************************************/
    diag_setup ( input_vars, para_vars, geom_vars, ierr, error );

    glmax_i ( ierr, COMM_SNAP );

    if ( *ierr != 0 )
    {
        tmpStrLen = strlen ("   ***ERROR: DIAG_SETUP:"
                            " Allocation error of diag type array.\n");
        ALLOC_STR( *error, tmpStrLen + 1, ierr);
        snprintf ( (char *) *error, tmpStrLen + 1,
                   "   ***ERROR: DIAG_SETUP:"
                   " Allocation error of diag type array.\n" );

        print_error ( fp_out, *error, IPROC, ROOT );

        stop_run ( 3, 3, 0, para_vars, sn_vars, data_vars, mms_vars,
                   geom_vars, solvar_vars, control_vars );
    }

    t2 = wtime ();

    TPARAM += t2 - t1;

/***********************************************************************
 * The time loop solves the problem for nsteps. If static, there is
 * only one step, and it does not have any time-absorption or -source
 * terms. Set the pointers to angular flux arrays. Set time to one for
 * static for proper multiplication in octsweep.
 ***********************************************************************/
// TODO: Add USEMKL and better parallelize functions
    if ( IPROC == ROOT )
    {
        fprintf ( fp_out,
                  "          Iteration Monitor\n"
                  "****************************************"
                  "****************************************\n" );
    }

    tot_iits = 0;

    // time_loop
    for ( cy = 1; cy <= NSTEPS; cy++ )
    {
        t3 = wtime();

        for ( i = 0; i < NG; i++ )
        {
            VDELT_1D(i) = 0;
        }
        time = 1;

        if ( TIMEDEP == 1 )
        {
            if ( IPROC == ROOT )
            {
                fprintf ( fp_out, " \n******************************\n\n"
                          "  Time Cycle %i\n", cy );
            }

            for ( i = 0; i < NG; i++ )
            {
                VDELT_1D(i) = 2 / (DT * V_1D(i));
            }
            time = DT * (((double) cy) - 0.5 );
        }

        if ( cy > 1 )
        {
            ptr_tmp = PTR_OUT;
            PTR_OUT = PTR_IN;
            PTR_IN = ptr_tmp;
        }

/***********************************************************************
 * Scale the manufactured source for time
 ***********************************************************************/
        if ( SRC_OPT == 3 )
        {
            if ( cy == 1 )
            {
                // TODO: Add USEMKL version
                for ( g = 0; g < NG; g++ )
                {
                    for ( oct = 0; oct < NOCT; oct++ )
                    {
                        for ( z = 0; z < NZ; z++ )
                        {
                            for ( y = 0; y < NY; y++ )
                            {
                                for ( x = 0; x < NX; x++ )
                                {
                                    for ( ang = 0; ang < NANG; ang++ )
                                    {
                                        QIM_6D(ang,x,y,z,oct,g) *= time;
                                    }
                                }
                            }
                        }
                    }
                }
            }

            else
            {
                sf = ((double) ( 2*cy - 1 )) / ((double) ( 2*cy - 3 ));
                // TODO: Add USEMKL version
                for ( g = 1; g <= NG; g++ )
                {
                    for ( oct = 0; oct < NOCT; oct++ )
                    {
                        for ( z = 0; z < NZ; z++ )
                        {
                            for ( y = 0; y < NY; y++ )
                            {
                                for ( x = 0; x < NX; x++ )
                                {
                                    for ( ang = 0; ang < NANG; ang++ )
                                    {
                                        QIM_6D(ang,x,y,z,oct,(g-1)) *= sf;
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }

/***********************************************************************
 * Zero out flux arrays. Use threads when available.
 ***********************************************************************/
#pragma omp parallel for schedule(dynamic, 1) default(shared) \
    private(g)
        for ( g = 1; g <= NG; g++ )
        {
            for ( z = 0; z < NZ; z++ )
            {
                for ( y = 0; y < NY; y++ )
                {
                    for ( x = 0; x < NX; x++ )
                    {
                        FLUX_4D(x,y,z,(g-1)) = 0;
                        for ( i = 0; i < (CMOM - 1); i++ )
                        {
                            FLUXM_5D(i,x,y,z,(g-1)) = 0;
                        }
                    }
                }
            }
        }

/***********************************************************************
 * Using Jacobi iterations in energy, and the work in the outer loop
 * will be parallelized with threads.
 ***********************************************************************/
        OTRDONE = false;

        cy_iits = 0;

        if ( (IPROC == ROOT) && (IT_DET == 0) ) fprintf(fp_out, "  Outer\n");

        t4 = wtime();

        TPARAM += t4 - t3;

        // outer_loop
        for ( otno = 1; otno <= OITM; otno++ )
        {
            t5 = wtime();

            if ( IPROC==ROOT && IT_DET==1 )
            {
                fprintf(fp_out, " ********************\n  Outer %i\n", otno);
            }

/***********************************************************************
 * Prepare some cross sections: total, in-group scattering, absorption.
 * Keep in the time loop for better consistency with PARTISN. Set up
 * geometric sweep parameters. Parallelize group loop with threads.
 ***********************************************************************/
#pragma omp parallel for schedule(dynamic, 1) default(shared)   \
    private(g)
            for ( g = 1; g <= NG; g++ )
            {
                expxs_reg  ( input_vars, data_vars, solvar_vars, NULL, g, 0, 0 );
                expxs_slgg ( input_vars, data_vars, solvar_vars, g );
                param_calc ( input_vars, sn_vars, solvar_vars, data_vars,
                             geom_vars, g );
            }

/***********************************************************************
 * Perform an outer iteration. Add up inners. Check convergence.
 ***********************************************************************/
            t6 = wtime();
            TPARAM += t6 - t5;

            out_iits = outer ( input_vars, para_vars, geom_vars, time_vars,
                               data_vars, sn_vars, control_vars, solvar_vars,
                               sweep_vars, dim_sweep_vars, fp_out, ierr );

            cy_iits += out_iits;

            if ( IPROC == ROOT )
            {
                if ( otno < 10 )
                {
                    fprintf ( fp_out, "    %i   Dfmxo= %.4E"
                              "    No. Inners=   %i\n", otno, DFMXO, out_iits );
                }
                else if ( otno < 100 )
                {
                    fprintf ( fp_out, "   %i   Dfmxo= %.4E"
                              "    No. Inners=   %i\n", otno, DFMXO, out_iits );
                }
                else if ( otno < 1000 )
                {
                    fprintf ( fp_out, "  %i   Dfmxo= %.4E"
                              "    No. Inners=   %i\n", otno, DFMXO, out_iits );
                }
                else
                {
                    fprintf ( fp_out, " %i   Dfmxo= %.4E"
                              "    No. Inners=   %i\n", otno, DFMXO, out_iits );
                }
            }

            if ( OTRDONE ) break;
        } // end outer_loop

/***********************************************************************
 * Print the time cycle details. Add time cycle iterations.
 ***********************************************************************/
        if ( TIMEDEP == 1 && IPROC == ROOT )
        {
            if ( OTRDONE )
            {
                fprintf ( fp_out, "\n  Cycle= %i"
                          "    Time= %.4E"
                          "    No. Outers= %i"
                          "    No. Inners= %i\n",
                          cy, time, otno, cy_iits );
            }
            else
            {
                fprintf ( fp_out, "\n  ***UNCONVERGED*** Stopping Iterations!!"
                          "\n  Cycle= %i"
                          "    Time= %.4E"
                          "    No. Outers= %i"
                          "    No. Inners= %i\n",
                          cy, time, otno, cy_iits );

            }
        }
        else if ( IPROC == ROOT )
        {
            if ( OTRDONE )
            {
                fprintf ( fp_out,
                          "\n  No. Outers= %i"
                          "    No. Inners= %i\n",
                          otno, cy_iits );
            }
            else
            {
                fprintf ( fp_out, "\n  ***UNCONVERGED*** Stopping Iterations!!"
                          "\n  No. Outers= %i"
                          "    No. Inners= %i\n",
                          otno, cy_iits );

            }
        }

        tot_iits += cy_iits;

        if ( !OTRDONE ) break;
    } // end time_loop


    if ( TIMEDEP == 1 && IPROC == ROOT )
    {
        fprintf ( fp_out, "\n******************************\n"
                  "  Total inners for all time steps, outers = %i\n", tot_iits );
    }

    if ( IPROC == ROOT )
    {
        fprintf ( fp_out, "\n****************************************"
                  "****************************************\n\n");
    }

    t7 = wtime();
    TSLV = t7 - t1;

    tmp = ((double) NX)   * ((double) NY_GL) * ((double) NZ_GL)
        * ((double) NANG) * ((double) NOCT)  * ((double) tot_iits);
    TGRIND = TSLV*1.0E9 / tmp;
}
